test/Concurrency/custom_executor_enqueue_availability.swift

```diff
@@ -0,0 +1,67 @@
+// RUN: %target-typecheck-verify-swift -enable-experimental-move-only
+// REQUIRES: concurrency
+// REQUIRES: OS=macosx
+
+/// Such a type may be encountered since Swift 5.5 (5.1 backdeployed) if someone implemented the
+/// not documented, but public Executor types back then already. Allow these to be implemented
+/// without warnings.
+@available(SwiftStdlib 5.1, *)
+final class OldExecutorOldStdlib: SerialExecutor {
+  func enqueue(_ job: UnownedJob) {}
+
+  func asUnownedSerialExecutor() -> UnownedSerialExecutor {
+    UnownedSerialExecutor(ordinary: self)
+  }
+}
+
+/// We warn on the ExecutorJob witness if the type has a broader
+/// availability, since in this case the UnownedJob version needs to exist.
+@available(SwiftStdlib 5.1, *)
+final class BothExecutorOldStdlib: SerialExecutor {
+  func enqueue(_ job: UnownedJob) {}
+
+  @available(SwiftStdlib 5.9, *)
+  func enqueue(_ job: __owned ExecutorJob) {} // expected-warning{{'Executor.enqueue(ExecutorJob)' will never be used, due to the presence of 'enqueue(UnownedJob)'}}
```

Comment:
What causes this to never be used? I would expect that if you're on a new-enough deployment target, both enqueues are available and can be used.

Reply:
They're both available on newer deployment targets but the impls delegate to each other by default

https://github.com/apple/swift/blob/300e7ce085af0ae050b54bd30cdf1d3eb8b9b500/stdlib/public/Concurrency/Executor.swift#L103-L120

with the `UnownedJob` impl ultimately being treated as the "source of truth"

https://github.com/apple/swift/blob/07f938435a8b7d15620a386d01f05c0094f04290/stdlib/public/BackDeployConcurrency/Executor.swift#L74-L78

This is also mentioned in the existing `custom_executor_enqueue_impls` unit test (which is similar to the one I added, albeit with availability checking disabled):

https://github.com/apple/swift/blob/07f938435a8b7d15620a386d01f05c0094f04290/test/Concurrency/custom_executor_enqueue_impls.swift#L22-L26

Reply:
Yes the reasoning here by @kabiroberai is correct -- we do this for compatibility reasons and the new warning does make sense.